File: core/api/src/main/java/com/iamkurtgoz/core/api/service/manager/ManagerService.kt
/*
 * Copyright 2024 Sporthor Android
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.iamkurtgoz.core.api.service.manager

import androidx.annotation.Keep
import com.iamkurtgoz.core.network.model.BaseRequest
import com.iamkurtgoz.core.network.model.BaseResponse
import com.iamkurtgoz.data.model.ClubResponseModel
import com.iamkurtgoz.data.model.GetSportClubResponseModel
import com.iamkurtgoz.domain.model.request.AddSportClubRequest
import com.iamkurtgoz.domain.model.request.DeleteCoachRequest
import com.iamkurtgoz.domain.model.request.UpdateCoachRequest
import com.iamkurtgoz.domain.model.request.UpdateSportClubFilesRequest
import com.iamkurtgoz.domain.model.request.UpdateSportClubRequest
import com.iamkurtgoz.domain.model.response.GetClubsAndDetailsDomainModel
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.POST

@Keep
interface ManagerService {

    @POST("Manager/AddSportClub")
    suspend fun addSportClub(@Body body: AddSportClubRequest?): Response<BaseResponse<ClubResponseModel>>

    @POST("Manager/UpdateSportClub")
    suspend fun updateSportClub(@Body body: UpdateSportClubRequest?): Response<BaseResponse<ClubResponseModel>>

    @POST("Manager/UpdateSportClubFiles")
    suspend fun updateSportClubFilesRequest(@Body body: UpdateSportClubFilesRequest?): Response<BaseResponse<ClubResponseModel>>

    @POST("Manager/GetSportClub")
    suspend fun getSportClub(): Response<BaseResponse<List<GetSportClubResponseModel>>>

    @POST("Manager/GetClubsAndDetails")
    suspend fun getClubsAndDetails(@Body body: BaseRequest = BaseRequest): Response<BaseResponse<GetClubsAndDetailsDomainModel>>

    @POST("Manager/DeleteCoach")
    suspend fun deleteCoach(@Body body: DeleteCoachRequest): Response<BaseResponse<Unit>>

    @POST("Manager/UpdateCoach")
    suspend fun updateCoach(@Body body: UpdateCoachRequest): Response<BaseResponse<Unit>>
}

File: Sporthor/Sporthor/Services/ManagerService.swift
//
//  ManagerService.swift
//  Sporthor
//
//  Created by derTurke on 14.06.2025.
//

import Foundation
import NetworkKit

enum ManagerService {
    case addSportClub(_ request: [String: Any])
    case updateSportClubFiles(_ request: [String: Any])
    case removeSportClubFiles(_ request: [String: Any])
    case getSportClub
    case updateSportClub(_ request: [String: Any])
    case removeSportClub(_ request: [String: Any])
    case getClubsAndDetails(_ request: [String: Any])
    case deleteCoach(_ request: [String: Any])
    case updateCoach(_ request: [String: Any])
}

extension ManagerService: NetworkService {
    var path: String {
        switch self {
        case .addSportClub:
            return "/api/Manager/AddSportClub"
        case .updateSportClubFiles:
            return "/api/Manager/UpdateSportClubFiles"
        case .removeSportClubFiles:
            return "/api/Manager/RemoveSportClubFiles"
        case .getSportClub:
            return "/api/Manager/GetSportClub"
        case .updateSportClub:
            return "/api/Manager/UpdateSportClub"
        case .removeSportClub:
            return "/api/Manager/RemoveSportClub"
        case .getClubsAndDetails:
            return "/api/Manager/GetClubsAndDetails"
        case .deleteCoach:
            return "/api/Manager/DeleteCoach"
        case .updateCoach:
            return "/api/Manager/UpdateCoach"
        }
    }
    var method: HTTPMethod {
        switch self {
        case .addSportClub,
                .updateSportClubFiles,
                .removeSportClubFiles,
                .getSportClub,
                .updateSportClub,
                .removeSportClub,
                .getClubsAndDetails,
                .deleteCoach,
                .updateCoach:
            return .POST
        }
    }
    
    var parameters: [String: Any]? {
        switch self {
        case .addSportClub(let request),
                .updateSportClubFiles(let request),
                .removeSportClubFiles(let request),
                .updateSportClub(let request),
                .removeSportClub(let request),
                .getClubsAndDetails(let request),
                .deleteCoach(let request),
                .updateCoach(let request):
            return request
        default:
            return nil
        }
    }
    
    var token: String? {
        return ApplicationContext.shared.authToken
    }
}
